### PR TITLE
[SPARK-42928][SQL] Make resolvePersistentFunction synchronized

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?

This PR makes the function `resolvePersistentFunctionInternal` synchronized.

### Why are the changes needed?

To make function resolution thread-safe.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing UTs.

### CODE REVIEW

**Comment by allisonwang-db** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 1]
```diff
@@ -1719,7 +1719,8 @@ class SessionCatalog(
       name: FunctionIdentifier,
       arguments: Seq[Expression],
       registry: FunctionRegistryBase[T],
-      createFunctionBuilder: CatalogFunction => FunctionRegistryBase[T]#FunctionBuilder): T = {
+      createFunctionBuilder: CatalogFunction => FunctionRegistryBase[T]#FunctionBuilder
+  ): T = synchronized {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1723]: which part is not thread-safe? `qualifyIdentifier`?

Comment (allisonwang-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1723]: Let's say we have two threads. If a persistent function has not been loaded into the function registry, the `if` condition here can return false for both threads.
```
if (registry.functionExists(qualifiedIdent)) {
  // This function has been already loaded into the function registry.
  registry.lookupFunction(qualifiedIdent, arguments)
} 
```
Then, if one thread successfully registers the persistent function in 
```
registerFunction(
    funcMetadata,
    overrideIfExists = false,
    registry = registry,
    functionBuilder = createFunctionBuilder(funcMetadata))
```
Then another thread can throw exceptions when trying to register this function again, due to `overrideIfExists = false`

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1723]: ah I see! Is this a long-standing bug?

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1723]: and can we put it as a code comment?

Comment (allisonwang-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala:1723]: > Is this a long-standing bug?

Yes, we should backport it

**Pull request closed without merging**